core-processor/src/precharge.rs

```diff
@@ -99,67 +108,52 @@ impl<'a> GasPrecharger<'a> {
         Ok(())
     }
 
-    pub fn charge_gas_for_program_data(
-        &mut self,
-        read_cost: u64,
-        per_byte_cost: u64,
-    ) -> Result<(), PrechargeError> {
-        self.charge_gas(
-            PreChargeGasOperation::ProgramData,
-            calculate_gas_for_program(read_cost, per_byte_cost),
-        )
+    pub fn charge_gas_for_program_data(&mut self) -> Result<(), PrechargeError> {
+        self.charge_gas(PreChargeGasOperation::ProgramData, self.read_cost.one())
     }
 
-    pub fn charge_gas_for_program_code_len(
-        &mut self,
-        read_cost: u64,
-    ) -> Result<(), PrechargeError> {
-        self.charge_gas(PreChargeGasOperation::ProgramCodeLen, read_cost)
+    pub fn charge_gas_for_program_code_len(&mut self) -> Result<(), PrechargeError> {
+        self.charge_gas(PreChargeGasOperation::ProgramCodeLen, self.read_cost.one())
     }
 
-    pub fn charge_gas_for_program_code(
-        &mut self,
-        read_cost: u64,
-        per_byte_cost: u64,
-        code_len_bytes: u32,
-    ) -> Result<(), PrechargeError> {
+    pub fn charge_gas_for_program_code(&mut self, code_len: Bytes) -> Result<(), PrechargeError> {
         self.charge_gas(
             PreChargeGasOperation::ProgramCode,
-            calculate_gas_for_code(read_cost, per_byte_cost, code_len_bytes.into()),
+            self.read_cost
+                .one()
+                .saturating_add(self.read_per_byte_cost.calc(code_len)),
         )
     }
 
-    pub fn charge_gas_for_instantiation(
-        &mut self,
-        gas_per_byte: u64,
-        code_length: u32,
-    ) -> Result<(), PrechargeError> {
-        let amount = gas_per_byte.saturating_mul(code_length as u64);
-        self.charge_gas(PreChargeGasOperation::ModuleInstantiation, amount)
+    pub fn charge_gas_for_instantiation(&mut self, code_len: Bytes) -> Result<(), PrechargeError> {
+        self.charge_gas(
+            PreChargeGasOperation::ModuleInstantiation,
+            self.instantiation_per_byte_cost.calc(code_len),
+        )
     }
 
     pub fn charge_gas_for_instrumentation(
         &mut self,
-        instrumentation_cost: u64,
-        instrumentation_byte_cost: u64,
-        original_code_len_bytes: u32,
+        original_code_len_bytes: Bytes,
     ) -> Result<(), PrechargeError> {
-        let amount = instrumentation_cost.saturating_add(
-            instrumentation_byte_cost.saturating_mul(original_code_len_bytes.into()),
-        );
-        self.charge_gas(PreChargeGasOperation::ModuleInstrumentation, amount)
+        self.charge_gas(
+            PreChargeGasOperation::ModuleInstrumentation,
+            self.instrumentation_cost.one().saturating_add(
+                self.instrumentation_cost_per_byte
+                    .calc(original_code_len_bytes),
+            ),
```

Comment:
We can introduce methods like
```rust
impl CostPer<Call> {
    pub fn bytes(per_byte: CostPer<Bytes>, amount: Bytes) -> u64 { ... }
}
```
To make it type-safe. Literally what `cost_with_per_byte!` does. Same for calculations with `reserve_for`.

Add TODO and open issue, please

Reply:
Done